Analysis/CBufferProcessor.3
'\"
'\" Copyright (c) 2004  Michigan State University
'\" The TCL formatting macros have been shamelessly stolen
'\" from the httpd tcl man page:
'\"
'\" Copyright (c) 1995-1997 Sun Microsystems, Inc.
'\" Copyright (c) 1998-2000 by Ajuba Solutions.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
'\" The definitions below are for supplemental macros used in Tcl/Tk
'\" manual entries.
'\"
'\" .AP type name in/out ?indent?
'\"	Start paragraph describing an argument to a library procedure.
'\"	type is type of argument (int, etc.), in/out is either "in", "out",
'\"	or "in/out" to describe whether procedure reads or modifies arg,
'\"	and indent is equivalent to second arg of .IP (shouldn't ever be
'\"	needed;  use .AS below instead)
'\"
'\" .AS ?type? ?name?
'\"	Give maximum sizes of arguments for setting tab stops.  Type and
'\"	name are examples of largest possible arguments that will be passed
'\"	to .AP later.  If args are omitted, default tab stops are used.
'\"
'\" .BS
'\"	Start box enclosure.  From here until next .BE, everything will be
'\"	enclosed in one large box.
'\"
'\" .BE
'\"	End of box enclosure.
'\"
'\" .CS
'\"	Begin code excerpt.
'\"
'\" .CE
'\"	End code excerpt.
'\"
'\" .VS ?version? ?br?
'\"	Begin vertical sidebar, for use in marking newly-changed parts
'\"	of man pages.  The first argument is ignored and used for recording
'\"	the version when the .VS was added, so that the sidebars can be
'\"	found and removed when they reach a certain age.  If another argument
'\"	is present, then a line break is forced before starting the sidebar.
'\"
'\" .VE
'\"	End of vertical sidebar.
'\"
'\" .DS
'\"	Begin an indented unfilled display.
'\"
'\" .DE
'\"	End of indented unfilled display.
'\"
'\" .SO
'\"	Start of list of standard options for a Tk widget.  The
'\"	options follow on successive lines, in four columns separated
'\"	by tabs.
'\"
'\" .SE
'\"	End of list of standard options for a Tk widget.
'\"
'\" .OP cmdName dbName dbClass
'\"	Start of description of a specific option.  cmdName gives the
'\"	option's name as specified in the class command, dbName gives
'\"	the option's name in the option database, and dbClass gives
'\"	the option's class in the option database.
'\"
'\" .UL arg1 arg2
'\"	Print arg1 underlined, then print arg2 normally.
'\"
'\" RCS: @(#) $Id$
'\"
'\"	# Set up traps and other miscellaneous stuff for Tcl/Tk man pages.
.if t .wh -1.3i ^B
.nr ^l \n(.l
.ad b
'\"	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ta \\n()Au \\n()Bu
.ie !"\\$3"" \{\
\&\\$1	\\fI\\$2\\fP	(\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
'\"	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
.AS Tcl_Interp Tcl_CreateInterp in/out
'\"	# BS - start boxed text
'\"	# ^y = starting y location
'\"	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
'\"	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
'\"	# VS - start vertical sidebar
'\"	# ^Y = starting y location
'\"	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.if !"\\$2"" .br
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
'\"	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
'\"	# Special macro to handle page bottom:  finish off current
'\"	# box/sidebar if in box/sidebar mode, then invoked standard
'\"	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
'\"	# DS - begin display
.de DS
.RS
.nf
.sp
..
'\"	# DE - end display
.de DE
.fi
.RE
.sp
..
'\"	# SO - start of list of standard options
.de SO
.SH "STANDARD OPTIONS"
.LP
.nf
.ta 5.5c 11c
.ft B
..
'\"	# SE - end of list of standard options
.de SE
.fi
.ft R
.LP
See the \\fBoptions\\fR manual entry for details on the standard options.
..
'\"	# OP - start of full description for a single option
.de OP
.LP
.nf
.ta 4c
Command-Line Name:	\\fB\\$1\\fR
Database Name:	\\fB\\$2\\fR
Database Class:	\\fB\\$3\\fR
.fi
.IP
..
'\"	# CS - begin code excerpt
.de CS
.RS
.nf
.ta .25i .5i .75i 1i
..
'\"	# CE - end code excerpt
.de CE
.fi
.RE
..
.de UL
\\$1\l'|0\(ul'\\$2
..
'\"  End of stolen macros --------------------------------
.TH "Buffer analysis framework library" 3 0.1 SpecTcl "NSCL Data Analysis packages"
.SH NAME
CBufferProcessor \- A callback based buffer processor for ad-hoc data analysis.
.SH SYNOPSIS
.DS
	/* Include.. */

#include <config.h>
#include <CBufferProcessor.h>

	/* To get: */

class CBufferProcessor
{

public:
  CBufferProcessor();
  virtual ~CBufferProcessor();
 

  bool          scalersSeen() const;
  unsigned int  scalerCount() const;
  unsigned long lastIncrement(unsigned int channel) const;
  float         Total(unsigned int channel) const;
  unsigned long lastIntervalStart() const;
  unsigned long lastIntervalEnd() const;
  std::string   Title() const;
  int           runNumber() const;
  std::string   runStartTime() const;
  std::string   runEndTime() const;

  void addCallback(unsigned int nBufferType, CBufferCallback& callback);
  void removeCallback(unsigned int nBufferType, CBufferCallback& callback);

  virtual void operator()(const void* pBuffer);

  static BufferTranslator*
                 getTranslatingPointer(const void* pBuffer);

};

	/* Somewhere in your code you need: */

void* gpTCLApplication;

	/* Link something like: */

  c++ -o yourprogram yourprogram.c -I$SpecTclHome/include  -L$SpecTclHome/lib \\
        -lAnalysis -ltclPlus -lException -lBufferAnalysis


.DE
.SH DESCRIPTION
.PP
 This class provides a callback based framework for analyzing files containing
NSCL formatted event data.  The idea is that the user will create and register
zero or more callback objects (See \fBCBufferCallback(3)\fR).  The data buffers
in the event file will then be pushed throug the buffer processor's function
call operator.  For each buffer, the processor will decode the buffer type and
dispatch to the function call operators of the set of buffer callbacks that
were registered for that type.
.PP
The processor registers a set of standard callbacks.  These standard callbacks,
invoked prior to any user defined callbacks, process state transition buffers
and scaler buffers to support the other member functions in the buffer
processing class.
.PP
If you are more comfortable programming in C than C++, see the
\fBBufferAnalysis(3)\fR man page for a C interface to the framework.
.SH PUBLIC INTERFACE
.TP
\fBbool scalersSeen() const\fR
This member function will return \fItrue\fR if the run has already processed at
least one scaler buffer.
.TP
\fBunsigned int  scalerCount() const\fR
This member function will return the number of scalers that are being read out
in the current run.  If \fBscalersSeen()\fR would not return \fItrue\fR, a
string exception will be thrown: ``No scaler buffers seen so I don't know how
many scalers there are''.
.TP
\fBunsigned long lastIncrement(unsigned int \fIchannel\fB) const\fR
Returns the most recent scaler increments read from channel number
\fIchannel\fR.  If \fBscalersSeen()\fR is \fIfalse\fR, this will throw the same
string exception as \fBscalerCount()\fR.  If the channel number does not select
a valid scaler channel, the string exception will be: ``Channel number out of
range''.
.TP
\fBfloat Total(unsigned int \fIchannel\fB) const\fR
Returns the current totals for the selected \fIchannel\fR.  If no scalers have
been seen, the same string exception as is thrown by \fBscalerCount()\fR will
be thrown.  If the channel selected is not valid, a string exception ``Invalid
channel to fetch totals'' will be thrown.
.TP
\fBunsigned long lastIntervalStart() const\fR
Returns the number of seconds of elapsed active running until the start of the
most recent scaler readout interval.  If scalers have not yet been processed
for this run, the string exception ``Scalers have not yet been seen, scaler
start interval invalid'' is thrown.
.TP
\fBunsigned long lastIntervalEnd() const\fR
Same as \fBlastIntervalStat()\fR, however the elapsed times returned are those
until the end of the most recent scaler read interval.
.TP
\fB std::string   Title() const\fR
Retrurns a string that contains the title of the current run.  If no buffers
with titles have been seen (note the start of run has a title), this will be an
empty string.  \fBSTD()\fR is a macro which, if necessary prefixes its
parameter with \fBstd::\fR. 
.TP
\fBint           runNumber() const\fR
returns the number of the run currently being processed.  This will be -1 if no
run has ever been processed.
.TP
\fBstd::string   runStartTime() const\fR
Returns the time at which this run started.  This will be a string of the form:
``Month day,year, hh:mm:ss'', for example ``February 18, 2005, 13:11:15''.
.TP
\fB std::string   runEndTime() const\fR
Once the run has been completely processed, this will return the time the run
ended as a timestamp string.  The form of the timestamp will be identical to
that returned by \fBrunStartTime()\fR.   Prior to the end of run, this will
return an empty string.
.TP
\fBvoid addCallback(unsigned int nBufferType, CBufferCallback& callback)\fR
Adds a callback to the processor.  \fInBufferType\fR is the type of buffer for
which the callback's function call operator should be invoked. \fIcallback\fR
is a reference to an object that is derived from the class
\fBCBufferCallback\fR (see \fBCBufferCallback(3)\fR).  Callbacks form an
ordered list for each buffer type.  A callback object can be registered on more
than one buffer type, however a callback object should be registered at most
once for a specific buffer type.
.TP
\fB void removeCallback(unsigned int nBufferType, CBufferCallback& callback)\fR
Removes a previously registerd callback from the buffer
processor. \fInBufferType\fR is the type of buffer on which this callback was
registered. \fIcallback\fR is a reference to the callback object that was
registered.    This member can throw string exceptions: If the buffer type does
not yet have any callback list created, the string exception will read:
``Buffer type has no callback lists''.  If the buffer type has a callback
list, but the specified object is not registered in it, the exception will
contain: ``No such callback in list''.
.TP
\fBvirtual void operator()(const void* \fIpBuffer\fB)\fR
Call this member to process a data buffer.   The buffer processor will decode
the buffer type and dispatch to any registered callbacks for that buffer type.
Callbacks will be called in the order in which they were registered.
.TP
\fB static BufferTranslator* getTranslatingPointer(const void* pBuffer)\fR
This static member function can be called to create a buffer translator object
appropriate to this buffer.  Buffer translators are used to instantiate
TranslatorBufferPointer objects, which provide a pointer like interface to
storage that does transparent byte swapping as required.  \fBNOTE\fR  the
pointer returned is dynamically allocated wth \fBnew\fR and must be freed by
the caller via \fBdelete\fR.
.SH EXAMPLE
.PP
Much of the error handling require for production software is omitted for this
example for the sake of brevity.
.PP
The example registers a callback object on all buffer types from 0 through
MAXBUFTYPE.  It counts the number of buffers of each type and, and the end of
the data file, prints out the histogram of buffer types with zeroes supressed.
.DS
#include <config.h>
#include <CBufferCallback.h>
#include <CBufferProcessor.h>
#include <buftypes.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include <iostream>

#ifdef HAVE_STD_NAMESPACE
using namespace std;
#endif

class HistogramCallback : public CBufferCallback
{
private:
  int histogram[MAXBUFTYPE];
public:
  void clear() {
    memset(histogram, 0, sizeof(histogram));
  }
  void dump(ostream& output) {
    for(int i=0; i < MAXBUFTYPE; i++) {
      if(histogram[i] != 0) {
	output << histogram[i] << " instances of buffer type: " << i << endl;
      }
    }
  }
  virtual void operator()(unsigned int ntype, const void* pBuffer) {
    histogram[ntype]++;
  }
};

int main(int argc, char** argv)
{
  unsigned short buffer[4096];
  int fd = open(argv[1], O_RDONLY);
  int nread;
  CBufferProcessor processor;
  HistogramCallback callback;

  for(int i=0; i < MAXBUFTYPE; i++) {
    processor.addCallback(i, callback);
  }

  callback.clear();
  
  while((nread = read(fd, buffer, sizeof(buffer))) > 0) {
    processor(buffer);
  }
  
  cout << "File : " << argv[1] << " run " << processor.runNumber()
       << " " << processor.Title() << endl;
  cout << "Run started: " << processor.runStartTime();
  cout << " Run ended: " << processor.runEndTime() << endl;
  cout << "Buffer type distribution: \n";

  callback.dump(cout);
}

void* gpTCLApplication;

.DE
.SH "SEE ALSO"
BufferAnalysis(3), CBufferCallback(3)

CHANGELOG
]
2.1-006 - Incorporate fix 113 - low level cutoffs lead to corrupted
          1-d displays.
        - Have build procedures clean any pre-built Gri prior to 
          tarring up the distribution.
2.1-007 - Incorporate fix 120 - memory leak on spectrum -delete -all
        - Date/time/version stamp installation.
2.1-008 
	- Incorporate fix for issue 95: parameter -list -byid fails to parse
          the output list for sorting.
        - Incorporate fix for issue 105: sbind usage level help refers to
          itself as the bind command not sbind.
        - Incorporate fix for issue 111: swrite creates a file even if
          no valid spectra are included on the command line.
2.1-009 
	- Incorporate printing fixes for last channel of 1-d.
	- Incorporate printinf fix ensure that 2-d spectrum channels display
      	  all non zeroes even if they are below the threshold for the lowest
	  color range.
        - Place upper limits on 1-d graphical entries at the top  of the
	  channel rather than at the bottom.  This more realistically
	  describes what's going on
        - Fix issue 118 Superpositions in window files are by id not name.
        - Fix issue 122 Speclist with a single spectrum gives blank output.
	- Fix issue 123 Print with superpositions crashes gri (gives it a
  	  bad command file).
	- Fix issue 124 FWHM not transformed to mapped coordinates.
2.1-010 - Fix Display/Makefile issues with lingering bison/flex output
          Add missing return value in ValidValue::operator=.
2.1-011 - Fix issue 135  swrite -format ascii does not retain the axis
             mapping information correctly.
        - Fallout from issue 133; fixed a defect in Display/axes.cc where
          computation of mapped channel tick mark intervals could cause
          DIV0 crashes in Xamine when the axis decreases from left to right. 
        - Backport colormanager intelligent visual choices from 2.2
2.1-012 - Backport fix for defect 144 from 2.2: Memory leak when user
          rejects events.
        - Backport repair for defect 149: Some cases where spectrum
          deletion from Xamine can kill both Xamine and SpecTcl
        - Backport repair for defect 145 Names with special characters
	  cause saveall to produce bad files.
2.1-013 - Backport filter generation fix from 2.2
          Repair defect 157: Graphical object copies fail if target
          is spectrum slot > 999
2.1-014
	- Fixed defect 151: printing 2-d's the channels have incorrect
          values, and log prints are just plain flat lined.
2.1-015 
        - Fix un-numbered defect channel command gets/sets wrong channel
          on Gamma 2-d's.  This was broken with the introduction of 
 	  arbitrary binning in 2.1.
Modifications for SpecTcl 2.2
January 30,2004  - Updated gri to gri-2.12.7
March 23, 2004   - Issue 97b - Add hierarchical unpacking capability to scripted spectcl.
                             - Also don't require begin run to pick up config.
March 31, 2004   - Added -rpath to contrib/scripted Makefile so that
                   LD_LIBRARY_PATH or ld.so.conf not needed.
April  5, 2004   - Fix issue 118 - low level cutoff display bug.
April 12, 2004   - Extract as many defs as possible from skeleton makefiles
                   to centrally located include files in order to keep the
 	           user makefiles constant.
                 - Use etc not Etc for etcetera stuff.
                 - Add buildstamp
April 13, 2004   - Fix memory leak on delete -all command: Spectrum
                   storage itself is not getting deleted.
                 - Fix lack of error checking on insufficient memory for
                   binding 2d spectrum (can segfault).
April 29, 2004   - Fix issue 95: parameter -list -byid fails.
April 29, 2004   - Fix issue 105: sbind Usage level help calls itself the
                   bind command which conflicts, of course, with the TCL
                   bind command.
April 29, 2004   - Fix issue 111 - swrite command creates a file even if no
                    existing spectra have been specified.
May 5, 2004      - Fix little diddly issues with printing and also
                   Put right limit at the top of a channel rather than the
                   left side of a channel.
                 - Fix ISSUE 118 superpositions in window files are written
	           as id's not as names.
	         - Fix issue 123 superpositions don't print
	         - Fix issue 124 FWHM is not translated to mapped coordinates
	           when spectrum is in mapped mode.
May 12, 2004     - Fix issue 49  
	           Fetching values from ValidValues that are in the
	           invalid state will now throw exceptions 
                 - Fix issue with Display/Makefile that clean leaves behind
                   detritus from flex and bison. (Added MOSTLYCLEANFILES 
                   to Display/Makefile.am).
June 8, 2004     - Increase spectrum size limits to 2^31.
June 17, 2004    - Defect 131 - There were problems generating the print tick
		   intervals when the axis specifications went in 'reverse'
		   (e.g. {100 -100 200}).  Fixed in modifications to
		   griprint.cc I think.
                 - Fix build issues in TestFiles directory so that the
                   event test generting programs can be built again Yeah.
July 1, 2004     - Defect 135 - swrite does not properly preserve the
                   axis mapping information.
	           Fallout from defect 133 discovered and fixed a defect
                   that causes the axis tick computation to DIV0 when
                   mapped axes have decreasing values going away from the
                   origin (e.g. {100 -100 201}).
July 2, 2004     - Defect 132 - Translator pointers are constrained to 
                   natural data boundaries.
                 - Defect 137 - Fix defect introducec by fix of Defect 132:
		   assumed getOffset was returning T offset it's now returning
		   byte offsets.

main/Analysis/CBufferProcessor.cpp
/*
    This software is Copyright by the Board of Trustees of Michigan
    State University (c) Copyright 2005.

    You may use this software under the terms of the GNU public license
    (GPL).  The terms of this license are described at:

     http://www.gnu.org/licenses/gpl.txt

*/
#include <config.h>
#include "CBufferProcessor.h"
#include "CBufferCallbackList.h"
#include "CBufferCallback.h"
#include "CStandardControlCallback.h"
#include "CStandardScalerCallback.h"
#include "SRunContext.h"

#include <TranslatorPointer.h>
#include <BufferTranslator.h>

#include <buftypes.h>
#include <buffer.h>

#ifdef HAVE_STD_NAMESPACE
using namespace std;
#endif


// Implements the CBufferProcessor class.  See the header for more information.

/*!
   Constructs a CBufferProcessor.  Wee will initialize the callback list
   with a set of empty lists To handle the MAXSYSBUFTYPE buffer number.  Note
   that the 0'th callback entry should be permanently empty as there is no buffer 
   type 0.
   We also create and initialize the run context block,
   and define/register the stock callbacks.
*/
CBufferProcessor::CBufferProcessor() :
  m_pRunContext(new SRunContext)
{
  // Extend the callbacks:

  extendCallbacks(MAXSYSBUFTYPE);

  // Initialize the context block:

  m_pRunContext->s_fScalersThisRun   = false;
  m_pRunContext->s_Title      = "";
  m_pRunContext->s_nRunNumber  = -1;
  m_pRunContext->s_RunStarted = "";
  m_pRunContext->s_RunEnded   = "";

  // Create and register the control and scaler callbacks:


  m_pControlCallback = new CStandardControlCallback(*m_pRunContext);
  addCallback(BEGRUNBF, *m_pControlCallback);
  addCallback(ENDRUNBF, *m_pControlCallback);
  addCallback(PAUSEBF,  *m_pControlCallback);
  addCallback(RESUMEBF, *m_pControlCallback);


  m_pScalerCallback = new CStandardScalerCallback(*m_pRunContext);
  addCallback(SCALERBF, *m_pScalerCallback);


}
/*!  
   Destroy a buffer processor.  The lists take care of themselves.
   We need to destroy the context block and our standard collection of buffer
   buffer processors:
*/
CBufferProcessor::~CBufferProcessor()
{
  delete m_pScalerCallback;
  delete m_pControlCallback;
  delete m_pRunContext;
}


// Selectors from the context block follow These prevent
// uncontrolled modification of the context block... we only hand out
// this block's pointer to standard callbacks we trust.

/*!
   Determine if a scaler buffer has been seen this run yet.
   \return bool
   \retval  true   - At least one scaler buffer has been seen.
   \retval  false  - No scaler buffers have ben seen.
*/
bool
CBufferProcessor::scalersSeen() const
{
  return m_pRunContext->s_fScalersThisRun;
}
/*!
   Return the number of scalers being read this run.
   If there have not yet been any scalers this run, we throw a string
   exception.  The string exception makes it a bit harsh on the C programmers.
   Oh well, maybe they should learn our language.

   \return unsigned int
   \retval number of scalers that are being read.

   \throw string  - if no scalers have been read yet.
*/
unsigned int
CBufferProcessor::scalerCount() const
{
  if(!scalersSeen()) {
    throw string("No scaler buffers seen so I don't know how many scalers there are");
  }
  return m_pRunContext->s_nScalerCount;
}
/*!
   Return a scaler channel's most recent increment.
   \param channel (unsigned int [in]):
      The number of the channel to fetch the increment for.  Channel numbers
      start from 0.

   \return unsigned long
   \retval the most recent increment value for the channel
   \throw string  
   - If the channel number is out of range.
   - If scalers have not yet been seen.
*/
unsigned long
CBufferProcessor::lastIncrement(unsigned int channel) const  
{
  if(channel < scalerCount()) {
    return m_pRunContext->s_Increments[channel];
  }
  else {
    throw string("Channel number out of range");
  }
}
/*!
  Return a scaler channel's current total.
  \param channel (unsigned int [in]):
      The number of the channel whose totals will be fetched.
      Channel numbers start from 0.
   \return float
   \retval The current total.
   \throw string
    - If the channel number is out of range.
    - If scalers have not been seen.
*/
float
CBufferProcessor::Total(unsigned int channel) const
{
  if (channel < scalerCount()) {
    return m_pRunContext->s_Totals[channel];
  } 
  else {
    throw string("Invalid channel to fetch totals");
  }
}
/*!
   Return the start time of the last scaler interval.
   \return unsigned long
   \retval    number of seconds since the start of the run.
   \throw string
   - If there have not yet been scaler buffers in this run.
*/
unsigned long
CBufferProcessor::lastIntervalStart() const
{
  if(m_pRunContext->s_fScalersThisRun) {
    return m_pRunContext->s_IntervalStartTime;
  }
  else {
    throw string("Scalers have not yet been seen, scaler start interval invalid");
  }
}
/*!
  Return the end time of the last scaler interval.
  \return unsigned long
  \retval number of seconds since the start of the run.
  \throw string
  - If there have been no scalers buffers for this run todate.
*/
unsigned long
CBufferProcessor::lastIntervalEnd() const
{
  if(m_pRunContext->s_fScalersThisRun) {
    return m_pRunContext->s_IntervalEndTime;
  }
  else {
    throw string("Scalers have not yet been seen, scaler start interval invalid");
  }
}
/*!
  Return the title of the most recent run.  If we have not seen any control buffers
  (possible for e.g. online processing of a run joined in progress), the title
  will be blank.
  \return string
  \retval the most recent title string seen.

*/
string
CBufferProcessor::Title() const
{
  return m_pRunContext->s_Title;
}
/*!
  Return the run number of the current run.  If control buffers have not yet
  been seen, this could be -1.
  \return int
  \retval most recent run number seen.
*/
int
CBufferProcessor::runNumber() const
{
  return m_pRunContext->s_nRunNumber;
}
/*!
  Get the absolute time at which the run started.  This is a string of 
  the form:

  month-name dd,yyyy hh:mm:ss

  e.g:

  February 11, 2005 11:03:15

  \return string
  \retval  Timestamp see above for format.
   - blank if not control buffers have been seen.

 */
string
CBufferProcessor::runStartTime() const
{
  return  m_pRunContext->s_RunStarted;
}
/*!
   Get the absolute time at which the most recent run ended.  Se runStartTime
   for more information about the return value of this function.
*/
string
CBufferProcessor::runEndTime() const
{
  return m_pRunContext->s_RunEnded;
}
/*!
   Add a callback for a buffer type to the list.
   \param nBufferType (unsigned int [in]):
       The type of buffer to add.
   \param callback (CBufferCallback& [in]):
       The callback object to add.
*/
void
CBufferProcessor::addCallback(unsigned int nBufferType, CBufferCallback& callback)
{
  extendCallbacks(nBufferType);	// Be sure there are enough entries and then:
  m_Callbacks[nBufferType].addCallback(callback); // Add to the list.

}
/*!
   Remove a callback given its reference, from a specific callback list.
   Since it's perfectly legitimate to have a callback object registered for more
   than one buffer type we force the user to specify which list to remove it from.

   \param nBufferType (unsigned int [in]):
      The type of buffer on which the callback was registered.
   \param callback (CBufferCallback& [in]):
      A reference to the callback that was added.

   \throw string 
      - nBufferType has no callback list associated with it.
      - The callback does not exist in the specified list.
*/
void
CBufferProcessor::removeCallback(unsigned int nBufferType, 
				 CBufferCallback& callback)
{
  if(nBufferType < m_Callbacks.size()) {
    CBufferCallbackList::CallbackIterator p = m_Callbacks[nBufferType].begin();
    CBufferCallbackList::CallbackIterator e = m_Callbacks[nBufferType].end();
    while(p != e) {
      if (*p == &callback) {
	m_Callbacks[nBufferType].removeCallback(p);
	return;
      }
      p++;
    }
    throw string("No such callback in list");
  }
  else {
    throw string("Buffer type has no callback lists");
  }
}
/*!
  Users should call this function to process a buffer.  
  - An appropriate translator pointer is created for the buffer.
  - The buffer type is extracted.
  - invokeCallbacks is called to process the buffer.

  Everything else is done by the callbacks (if any).
  \param pBuffer (const void* [in]):
    Pointer to the buffer.
*/
void
CBufferProcessor::operator()(const void* pBuffer)
{
  BufferTranslator* bt = getTranslatingPointer(pBuffer);
  TranslatorPointer<unsigned short> p(*bt);
  unsigned int nBufferType = p[1];


  invokeCallbacks(nBufferType, pBuffer);
  delete bt;
}

//     Utility(ies) for buffer callbacks:

/*!
   Create a translator pointer for a buffer.
   This TranslatorPointer<unsigned short> can be used to get byte order
   independent access to the contents of an NSCL data buffer.
   The user must free this pointer when done.
   \param pBuffer (const void*):
       A pointer to an NSCL data acquisition system buffer.
   \return TranslatorPointer<unsigned short>
   \retval a translating pointer to the first word of the buffer.
*/
BufferTranslator*
CBufferProcessor::getTranslatingPointer(const void* pBuffer)
{
  bheader *pHeader = (bheader*)pBuffer;

  return BufferFactory::CreateBuffer(pHeader,
				       pHeader->lsignature);
}

//   -- private utility functions these by defintion don't merit doxygenized
//      headers.

/*
   Invoke the callbacks associated with a specific buffer type:
   Parameters:
     nBufferType (unsigned int [in]):
           The buffer type for which we want to callback.
     pBuffer (const void* [in]):
           The buffer itself.
*/
void
CBufferProcessor::invokeCallbacks(unsigned int nBufferType,
				  const void*  pBuffer)
{
  if(nBufferType < m_Callbacks.size()) {
    m_Callbacks[nBufferType](nBufferType, pBuffer); // This iterates the list.
  }
}
/*
   Extend the callback list vector as needed so that there will be an entry
   for nBufferType
   \param nBufferType (unsigned int [in]):
      The buffer type for which we want a callback.
*/
void
CBufferProcessor::extendCallbacks(unsigned int nBufferType)
{
  int nType = m_Callbacks.size();
  while(m_Callbacks.size() <=  nBufferType) {
    CBufferCallbackList temp(nType);
    m_Callbacks.push_back(temp);	// This will be copy constructed in.
    nType++;
  }
}
